buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java

```diff
@@ -515,46 +538,98 @@ public void freeThreadLocalCache() {
+    private final class PoolThreadLocalCache {
+        private final FastThreadLocal<PoolThreadCache> threadLocalCache;
+
+        PoolThreadLocalCache(final boolean useCacheForAllThreads, boolean useThreadLocal) {
+            if (useThreadLocal) {
+                threadLocalCache = new FastThreadLocal<PoolThreadCache>() {
+                    @Override
+                    protected synchronized PoolThreadCache initialValue() {
+                        final PoolArena<byte[]> heapArena = leastUsedArena(heapArenas);
+                        final PoolArena<ByteBuffer> directArena = leastUsedArena(directArenas);
+                        final Thread current = Thread.currentThread();
+                        final EventExecutor executor = ThreadExecutorMap.currentExecutor();
+                        if (useCacheForAllThreads ||
+                                // If the current thread is a FastThreadLocalThread we will always use the cache
+                                current instanceof FastThreadLocalThread ||
+                                // The Thread is used by an EventExecutor,
+                                // let's use the cache as the chances are good that we will allocate a lot!
+                                executor != null) {
+                            final PoolThreadCache cache = new PoolThreadCache(
+                                    heapArena, directArena, smallCacheSize, normalCacheSize,
+                                    DEFAULT_MAX_CACHED_BUFFER_CAPACITY, DEFAULT_CACHE_TRIM_INTERVAL,
+                                    useCacheFinalizers(current));
+
+                            if (DEFAULT_CACHE_TRIM_INTERVAL_MILLIS > 0) {
+                                if (executor != null) {
+                                    executor.scheduleAtFixedRate(trimTask, DEFAULT_CACHE_TRIM_INTERVAL_MILLIS,
+                                            DEFAULT_CACHE_TRIM_INTERVAL_MILLIS, TimeUnit.MILLISECONDS);
+                                }
+                            }
+                            return cache;
+                        }
+                        // No caching so just use 0 as sizes.
+                        return new PoolThreadCache(heapArena, directArena, 0, 0,
+                                0, 0, false);
+                    }
+                    @Override
+                    protected void onRemoval(PoolThreadCache threadCache) {
+                        threadCache.free(false);
+                    }
+                };
+            } else {
+                threadLocalCache = null;
+            }
+        }
 
-        PoolThreadLocalCache(boolean useCacheForAllThreads) {
-            this.useCacheForAllThreads = useCacheForAllThreads;
+        PoolThreadCache get() {
+            if (threadLocalCache != null) {
+                return threadLocalCache.get();
+            }
+            return PoolThreadCache.THREAD_CACHE_WITHOUT_THREAD_LOCAL;
```

Comment:
The problem of using thread local is based on the caller - not a global behaviour IMO: you want a way to instruct the allocator to...
1. use thread local on the usual known platform thread type (configurable)
2. not use thread local if not

Reply:
Yes, maybe this can be done by `Thread.isVirtual()`? but currently it use the constructor param to decide whether 
disable threadLocal or not, also I wonder how `AdaptivePoolingAllocator` deal with this?

Reply:
I would instead provide some API to inject/override (not always possible to not break the public APIs) a different behaviour, instead.

> also I wonder how AdaptivePoolingAllocator deal with this?

the adaptive allocator never block and not-shareable (which doesn't belong to event loops) magazine are using an mpsc recycler for ByteBuf wrappers.
I've implemented something similar for the non adaptive allocator some time ago, but TBH the adaptive IMO is the way to go.
Maintaining a jemalloc-java version has been a PITA (no pun intended) along these year - ask @normanmaurer and @chrisvest about it...

Reply:
:smile:, Maybe it's true, but, IMHO, if the `AdaptivePoolingAllocator` wants to better deal with the memory-reusing and fragment-reduction issues, it will eventually end with the similar approaches which `PooledByteBufAllocator` has already done...

Reply:
Not sure, think about garbage collectors: the approach taken by the adaptive allocator is very very similar to OpenJDK TLABs which has served pretty well.
Clearly it work better if the allocator is used for networking and in the event loop, but the price for simplicity is to be biased specific use cases, which is a pros IMO.

Reply:
> Clearly it work better if the allocator is used for networking and in the event loop, but the price for simplicity is to be biased specific use cases, which is a pros IMO.

Yes, but to cover more different use cases, this RP gives the users one more option.